### ISSUE TITLE
High precision mode for location updates

### ISSUE THREAD

**Issue by Tho85** [home-assistant/android#973]
**Is your feature request related to a problem? Please describe.**
I would like to obtain more accurate location updates for a short period of time, in order to trigger time- and location-critical events. For example, I would like to open my garage as soon I approach by bike.

**Describe the solution you'd like**
I would like to have a high-precision location update mode that I could enable or disable through a special message from HA. In this mode, the app would send GPS-only location updates every 15 seconds. So, I would expect the following:

1. Send a notification to the app to start high precision mode

    ```yaml
    automations:
      - ...
        action:
          - service: notify.mobile_app
            data:
              message: start_location_tracking
    ```

2. The app would show a persistent notification "Home Assistant is tracking your location". A background service (?) continuously acquires GPS fixes every 15 seconds, and the existing location tracking mechanism sends these fixes to HA. Similar to fitness trackers.

3. Send a notification to the app to stop high precision mode

    ```yaml
    automations:
      - ...
        action:
          - service: notify.mobile_app
            data:
              message: stop_location_tracking
    ```
   Alternatively, it times out after e.g. 15 minutes.

**Describe alternatives you've considered, if any**

* *Background location*: Location provided by background location is not exact enough, as it is mostly based on Wifi or cell towers in my case. Additionally, it updates at most once a minute, so it triggers too late for my use case, if at all.
* *Zone background / location*: Same as above, it does not update fast enough for my case.
* *`request_location_update`*: Is not exact enough, as it uses the same location mechanism as background location (i.e. Wifi or cell tower).
* *Other triggers like Wifi*: Triggering too late. And it also works for locations where there is Wifi reception. This would not work for locations outside home.

**Additional context**
In my use case, I would add a broader zone around my garage (e.g. 300m in diameter) that would start high precision tracking. When I enter a narrower zone (e.g. 50m in diameter), it would open my garage and stop high precision tracking.

When I have a fitness tracker running in parallel, HA also receives the location update intents. So in order to implement this, we would only need a background service and a notification, and the location update mechanism could stay as it is. This would also benefit other use cases like triggering geofence zones.

**Comment by dshokouhi** [comment]
So one new feature in 2.5.0 is the ability to set faster updates via a sensor setting for the Single Accurate Location sensor.  I know its not what you are asking for but currently there is no way for the app to receive commands to perform actions from HA.  This might help make the updates faster for you by adjusting it.  The current interval is set to 60 seconds.  Keep in mind as well that even with adjusting this if your accuracy did not meet the requirement, location still won't be sent so you may need to adjust that accuracy setting as well.

https://companion.home-assistant.io/docs/core/location#android-location-settings

**Comment by Tho85** [comment]
> So one new feature in 2.5.0 is the ability to set faster updates via a sensor setting for the Single Accurate Location sensor. [...] This might help make the updates faster for you by adjusting it. The current interval is set to 60 seconds. Keep in mind as well that even with adjusting this if your accuracy did not meet the requirement, location still won't be sent so you may need to adjust that accuracy setting as well.

Yes, I've seen that change in the Git history, but I guess it won't help much with accuracy, as you pointed out. It looks like it is already in the Beta version, so I'll try it.

> I know its not what you are asking for but currently there is no way for the app to receive commands to perform actions from HA.

What about the `request_location_update` command? From my understanding this is a special command sent from HA, to which the app reacts. Couldn't we use something similar to start this "high precision" mode on the device?

**Comment by dshokouhi** [comment]
> What about the `request_location_update` command? From my understanding this is a special command sent from HA, to which the app reacts. Couldn't we use something similar to start this "high precision" mode on the device?

You can just use that command today to request updates as often as you need.  That command uses the Single Accurate Location sensor settings so once you get that command to work as you expect just use it in your automation.  Maybe tie it to an input boolean so you can turn it on and off via a widget?  This way you can build your own high precision mode to meet your own accurate and timing needs.  We really made the command more flexible with the setting change.

**Comment by dshokouhi** [comment]
Have you also considered using this integration? https://www.home-assistant.io/integrations/proximity/

**Comment by Tho85** [comment]
> You can just use that command today to request updates as often as you need.

I already did that. I created an automation that would request single location updates every x seconds, for y minutes. This didn't work, probably due to the timing limitations inherent to `request_location_updates` as stated [in the docs](https://companion.home-assistant.io/docs/notifications/notification-commands/#request-location-updates). Location updates would trigger far too late (e.g. yesterday I received a "zone entered" event more than five minutes after entering a zone, with "minimum time between updates" set to 15000).

The proximity integration wouldn't be of much use. My daily route only takes about four minutes, and three minutes into that I'd like to open the garage. If HA doesn't see my location within the first three minutes (see above), there's not much that proximity can do.

The app's current approach to location tracking resembles OwnTracks's [_significant location change mode_](https://owntracks.org/booklet/features/location/), while I'm thinking about something like their _move mode_. A workaround would be to activate OwnTracks's move mode using an [intent](https://owntracks.org/booklet/features/android/#automation-via-tasker-automagic-etc) and have Home Assistant receive the location updates more often. But right now the intent needs to have an extra field set, which the Home Assistant app doesn't yet support.
